src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -822,17 +822,65 @@ internal override TypeSymbol SetUnknownNullabilityForReferenceTypes()
                 }
             }
 
-            if (!haveChanges)
+            TypeSymbol result = this;
+            if (haveChanges)
             {
-                allTypeArguments.Free();
-                return this;
+                var definition = this.OriginalDefinition;
+                TypeMap substitution = new TypeMap(definition.GetAllTypeParameters(), allTypeArguments.ToImmutable());
+                result = substitution.SubstituteNamedType(definition);
             }
-            else
+
+            allTypeArguments.Free();
+            return result;
+        }
+
+        internal override bool TryMergeNullability(TypeSymbol other, out TypeSymbol merged)
+        {
+            merged = this;
+
+            if (!IsGenericType)
             {
-                TypeMap substitution = new TypeMap(this.OriginalDefinition.GetAllTypeParameters(),
-                                                   allTypeArguments.ToImmutableAndFree());
+                return true;
+            }
 
-                return substitution.SubstituteNamedType(this.OriginalDefinition);
+            var definition = this.OriginalDefinition;
+            var allTypeParameters = definition.GetAllTypeParameters();
+            var allTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance();
+            GetAllTypeArgumentsNoUseSiteDiagnostics(allTypeArguments);
+            var otherTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance();
+            ((NamedTypeSymbol)other).GetAllTypeArgumentsNoUseSiteDiagnostics(otherTypeArguments);
+
+            try
+            {
+                bool haveChanges = false;
+                for (int i = 0; i < allTypeArguments.Count; i++)
+                {
+                    TypeSymbolWithAnnotations oldTypeArgument = allTypeArguments[i];
+                    TypeSymbolWithAnnotations newTypeArgument;
+                    if (!oldTypeArgument.TryMergeNullability(otherTypeArguments[i], allTypeParameters[i].Variance, out newTypeArgument))
```

Comment:
>allTypeParameters[i].Variance [](start = 84, length = 29)

This is not correct. We need the variance from "above". If the variance from above is invariant, that is what we need to use here. If the variance from above is covariant, then this is correct. If the variance from above is contravariant, then we need to swap the roles of "in" and "out".